cmd/joy/release.go

```diff
@@ -326,82 +330,32 @@ func NewGitCommands() *cobra.Command {
-				}
-
-				repoDir := filepath.Join(sourceDir, path.Base(repository))
-				if _, err := os.Stat(repoDir); err != nil {
-					if !errors.Is(err, os.ErrNotExist) {
-						return err
-					}
-
-					cloneOptions := github.CloneOptions{
-						RepoURL: repository,
-						OutDir:  repoDir,
-					}
-					if err := github.Clone(sourceDir, cloneOptions); err != nil {
-						return fmt.Errorf("failed to clone project: %w", err)
-					}
-				}
-
-				if err := git.FetchTags(repoDir); err != nil {
-					return fmt.Errorf("fetching git tags: %w", err)
-				}
-
-				tmpl, err := func() (*template.Template, error) {
-					templateSource := cmp.Or(project.Spec.GitTagTemplate, cfg.DefaultGitTagTemplate)
-					if templateSource == "" {
-						return nil, nil
-					}
-					return template.New("").Parse(templateSource)
-				}()
+				repoDir, err := project.GetCachedSourceDir(sourceRelease.Project, cfg.GitHubOrganization, cfg.RepositoriesDir, cfg.JoyCache)
 				if err != nil {
-					return fmt.Errorf("parsing config gitTagTemplate: %w", err)
-				}
-
-				getRevision := func(release *v1alpha1.Release) (string, error) {
-					if tmpl == nil {
-						return release.Spec.Version, nil
-					}
-
-					var buffer bytes.Buffer
-					if err := tmpl.Execute(&buffer, struct{ Release *v1alpha1.Release }{release}); err != nil {
-						return "", fmt.Errorf("executing template: %w", err)
-					}
-
-					return buffer.String(), nil
+					return fmt.Errorf("cloning repository: %w", err)
 				}
 
-				sourceTag, err := getRevision(sourceRelease)
+				sourceTag, err := release.GetGitTag(sourceRelease, cfg.DefaultGitTagTemplate)
 				if err != nil {
-					return fmt.Errorf("getting source tag from release: %w", err)
+					return fmt.Errorf("getting tag for source version %s of release %s: %w", sourceRelease.Spec.Version, sourceRelease.Name, err)
 				}
 
-				targetTag, err := getRevision(targetRelease)
+				targetTag, err := release.GetGitTag(targetRelease, cfg.DefaultGitTagTemplate)
 				if err != nil {
-					return fmt.Errorf("getting target tag from release: %w", err)
+					return fmt.Errorf("getting tag for target version %s of release %s: %w", targetRelease.Spec.Version, targetRelease.Name, err)
 				}
 
 				gitArgs := append([]string{"git", command}, args[1:]...)
 				gitArgs = append(gitArgs, sourceTag+".."+targetTag)
 
-				if project.Spec.RepositorySubpaths != nil {
+				if sourceRelease.Project.Spec.RepositorySubpaths != nil {
 					if !slices.Contains(gitArgs, "--") {
 						gitArgs = append(gitArgs, "--")
 					}
-					gitArgs = append(gitArgs, project.Spec.RepositorySubpaths...)
+					gitArgs = append(gitArgs, sourceRelease.Project.Spec.RepositorySubpaths...)
 				}
 
-				fmt.Println(color.InCyan("running:"), color.InBold(strings.Join(gitArgs, " ")), color.InCyan("in"), color.InBold(repoDir))
+				fmt.Println(color.InCyan("running:"), color.InBold("git "+strings.Join(gitArgs, " ")), color.InCyan("in"), color.InBold(repoDir))
```

Comment:
"git" was already included in the args. this will print: git git, which I've heard is rude in England.

Reply:
I'd rather say that what's a bit rude is naming a variable `gitArgs` and... putting more than just args in it, just to make the print easier, to then later remove the extra stuff when you really just needs the args part! 😜  See, it even got myself confused as to what to expect to have in that variable...